File: kangaroo_app_sdk_android/android/src/main/kotlin/com/kangaroo/flutterplugin/android/features/depositGiftCard/GiftCardDepositHandler.kt
package com.kangaroo.flutterplugin.android.features.depositGiftCard

import com.kangaroo.flutterplugin.android.base.PluginChannelHandler
import com.kangaroo.flutterplugin.android.base.pushSerializedResultToEventSink
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import kangaroorewards.appsdk.features.depositGiftCard.GiftCardDepositApi
import kangaroorewards.appsdk.features.depositGiftCard.serializeGiftCardDepositState
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json


class GiftCardDepositHandler : EventChannel.StreamHandler, PluginChannelHandler {
    var sink: EventChannel.EventSink? = null

    override val methodChannel: String
        get() = "customer_sdk/methods/deposit_gift_card"

    override val eventChannel: String
        get() = "customer_sdk/events/deposit_gift_card"

    override fun onMethodCall(call: MethodCall): Unit? {
        return depositGiftCard(call)
    }

    override fun getStreamHandler(): EventChannel.StreamHandler {
        return this
    }

    companion object {
        fun depositGiftCard(call: MethodCall): Unit? {
            GiftCardDepositApi().depositGiftCard(
                depositId = call.argument<String>("depositId") as String
            )
            return null
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        GiftCardDepositApi().giftCardDepositState.serializeGiftCardDepositState().observe { result ->
            pushSerializedResultToEventSink<GiftCardDepositApi>(sink, result)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}

File: kangaroo_app_sdk_ios/ios/Classes/features/depositGiftCard/GiftCardDepositHandler.swift
import Foundation
import Flutter
import KangarooAppSdkCustomer

class GiftCardDepositHandler: NSObject, FlutterStreamHandler, PluginChannelHandler {
    var sink: FlutterEventSink?

    var methodChannel: String = "customer_sdk/methods/deposit_gift_card"

    var eventChannel: String = "customer_sdk/events/deposit_gift_card"

    func onMethodCall(call: FlutterMethodCall) async -> Any? {
        return await GiftCardDepositHandler.depositGiftCard(call: call)
    }

    func getStreamHandler() -> (FlutterStreamHandler & NSObjectProtocol)? {
        return self
    }


    static func depositGiftCard(call: FlutterMethodCall) async -> String? {
        

        


        

        guard let args = call.arguments else {
            return nil
        }
        do {
       if let myArgs = args as? [String: Any] {
                        guard let depositId = myArgs["depositId"] as? String else {return nil}

         let result = try await GiftCardDepositApi().depositGiftCard(
                depositId: depositId
            ).serializeGiftCardDepositApiResult()

        switch result {
            case let result as SerializedResultSuccess:
                return result.data
            case let result as SerializedResultUnauthorizedError:
                return result.error
            case let result as SerializedResultUnknownError:
                return result.error
            default:
                return nil
                }
            }
        }
        catch {
            return nil
        }
        
        return nil
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping
        FlutterEventSink) -> FlutterError? {
        sink = events
        GiftCardDepositApi().giftCardDepositState
            .serializeGiftCardDepositState().observe { result in
            switch result {
            case let result as SerializedResultIdle:
                self.sink?(result.state)
            case let result as SerializedResultLoading:
                self.sink?(result.state)
            case let result as SerializedResultSuccess:
                self.sink?(result.data)
            case let result as SerializedResultEmptyResponse:
                self.sink?(result.body)
            case let result as SerializedResultUnauthorizedError:
                self.sink?(result.error)
            case let result as SerializedResultUnknownError:
                self.sink?(result.error)
            default:
                break
            }
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}
